### PR TITLE
Enhancement: Do not slow down external perimeters to meet minimum layer time

### LINKED ISSUE #5147: Don't slow down external walls when enforcing minimum layer time
### Is there an existing issue for this feature request?

- [X] I have searched the existing issues

### Is your feature request related to a problem?

Printer slows down, external wall glossiness/ringing pattern/corner sharpness changes, making prints ugly.

### Which printers will be beneficial to this feature?

All

### Describe the solution you'd like

Slow down internal parts only and don't slow down outer walls.

### Describe alternatives you've considered

Use a layer height modifier to slow down inner parts manually on fast layers.

### Additional context

_No response_

### LINKED ISSUE #5573: Minimum layer time: decrease acceleration before slowing down print speed
### Is there an existing issue for this feature request?

- [X] I have searched the existing issues

### Is your feature request related to a problem?

Acceleration impacts print quality, especially in modern printers which push input shaping too much and get quite softer corners.

When layers are too fast, some choice has to be made about slowing them down.

### Which printers will be beneficial to this feature?

All

### Describe the solution you'd like

I would offer the user the option to reduce acceleration first, down to a minimum value, and then reduce print speed. 

This way, at least print quality and details improve if a layer has to be slowed down.

### Describe alternatives you've considered

Potential issue is only that print quality varies along the height, but the same applies when speed is reduced so my proposal is not intrinsically worse, just a different compromise.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by igiannakas**
_Ready for review - PR description updated 5th June '24._

This PR aims to allow the user to skip adjusting the speed of external perimeters to meet the minimum layer time threshold.

### **Problem statement:**
When printing gloss like material, variations in external wall speed can cause significant change in wall shine, as the filament has less time to melt in the extruder. 

The external wall speed variation can be caused by several factors:
1. Slowdown for overhangs (the old classic version). By using the non classic version this is mitigated
2. Slowdown of the entire layer to meet layer time targets 

The artefact can be seen below. PLA print, at 220C where some areas are printed faster than others, causing a variation in gloss.

IMG_3871

Equally, having uneven external wall print speeds between layers can cause uneven extrusion as the extruder flow rate is not always ideal and it tends to under extrude the faster you print. In addition it can cause uneven material shrinkage resulting in what appears like "z banding". 

### **What this PR does:**

This PR enables the user to opt out of slowing down the external perimeter to meet minimum layer time targets. The outcome is a consistent external wall print speed (excluding slowdowns for overhangs) which should materially improve surface finish.

This option is available per filament as some filaments may benefit from this more than others and is consistent with the current convention of defining cooling parameters on a per filament basis.

image

### Results:

**Before:**
image
(notice how the speed variations caused by the minimum layer time cause the external wall surface to create "bands". These may be visible as z banding in the final print. these are caused by the slight difference in time each internal sparse infill takes to print)

**After:**
image

Notice how the external walls are not slowed down but the internal features are. 

**Before:**
image

**After:**
image

**It is also compatible with multi material printing:**
The white filament is with this setting on.
image
The white filament external perimeter speed is not slowed down
image
image

**The feature respects slowdown for overhangs and extrusion rate smoothing as seen below:**
image

Fixes #5147 
Fixes #5573

### CODE REVIEW

**Comment by celtare21** [comment]
Really hope to see this integrated!

**Comment by Ataraxia-Mechanica** [comment]
Currently does not seem to be working. Can't wait for this to be a thing.

**Comment by igiannakas** [comment]
> Currently does not seem to be working. Can't wait for this to be a thing.

New version is up, waiting for the build

**Comment by ramosglauco** [comment]
Big idea! I hope it is implemented soon.

**Comment by Ataraxia-Mechanica** [comment]
No idea why nobody came up with this idea till now, and why I didn't come up with this till I modded my 3d printer to print so fast that almost all things I print have to slow down due to cooling XD.

**Comment by Ataraxia-Mechanica** [comment]
Found one case where this would be extremely useful. When printing with "infill combination" on, the layers with infill would  take significantly more time than non-infill layers, causing uneven outer surface. Printing the test model right now.

**Comment by Ataraxia-Mechanica** [comment]
In this case, the outer wall in 2.0.0 would slow down just enough to hit the VFA speed of my printer causing resonance, while with no outer wall slowdown POC, the outer surface is much smoother.
Tested with PETG at high temperature, with belts overtightened to show the artifacts as much as possible.

[girdfinity-1x1x6-voron.zip](https://github.com/SoftFever/OrcaSlicer/files/15136989/girdfinity-1x1x6-voron.zip)

==========================================

slowdown
photo 1

==========================================

no slowdown
photo 2

**Comment by igiannakas** [comment]
Thank you!

**Comment by SoftFever** [comment]
Wow!
This is a fantastic idea!

**Comment by igiannakas** [comment]
I’ll spend a bit of time working this feature as it’s by far incomplete right now, to get it to a better testing state ;)

**Comment by StuSerious** [comment]
@igiannakas let me know if you need some testing like I did for https://github.com/SoftFever/OrcaSlicer/pull/4333 ;)

**Comment by igiannakas** [comment]
> @igiannakas let me know if you need some testing like I did for #4333 ;)

Yes please - testing requirements are in the top post together with limitations. This PR is not ready for daily use and only in a controlled environment as there are many things that don’t work yet - trying to demonstrate value and then I can devote time to productionise it.

**Comment by Zibit-Meowe** [comment]
Just want to add I've been using this extensively for the last couple weeks.  It's been amazing.

I've been getting the best large ASA parts I've ever printed.  So far I haven't seen any strange behavior from slowing down for overhangs.  But with ASA, my overhangs aren't a ton slower than the external walls anyway unless basically bridging.

I've only had one issue that I'm testing.  With thin [basically 6 layer widths thick or less] walls that can't hit minimum time unless outer wall is slowed, and scarf around entire wall, there seems to be some strange behavior.

I don't want to speculate too much.  I need to ensure it's something else.  Probably just a scarf or general slicing bug.

I'll check against latest live now that I have a model that seems to show this more often than not.  If it does end up related, I'll post some pictures.

Other that whatever that issue is, this has been the biggest increase in ABS/ASA print quality from any feature I've seen in a long time.

**Comment by igiannakas** [comment]
Gotcha thank you for the testing. I’ll proceed with making it operational - you should really not be using this build for “production” purposes as it barely works and has a ton of bugs :)

**Comment by igiannakas** [comment]
Also would you be able to post any images?

**Comment by Zibit-Meowe** [comment]
Sorry, was on the road for work.  Will be back printing again tonight.  I'll try to get a couple good screenshots when I get home  from work.

**Comment by Zibit-Meowe** [comment]
Alright.   It's late, so apologies for there not being more pictures yet.  I tried my best to get lighting so you can see what I see looking at it.

Here are the good results....  All Inner Outer Inner.  Some have extrustion rate smoothing.  Some have scarf around all wall.

ASA/ABS
I have lots of different parts of these.   They all look this good.  The gray is the easiest to tell just how amazing the consistency is of the outer wall.  It's a hard part to keeps consistent finish at high temp.  These are all at 265 to 270C, so without this they have a lot of banding.  I'll talk more about the ammo boxes in the PETG pics.










PETG
The Blue one was from tonight, using latest 2.1 Beta [not nightly] and it's hard to see, but if I print one that keeps gloss consistency, I outrun that filament's ability to flow well and get issues on the internal walls and .  Could be that filament [overture petg] so I'm drying it for another 24 hours and we'll see what it looks like after.  Regardless, getting the perfect finishes is much harder normally.


You can see a minor layer shrink with the right angle of the flash light.  But that's it.

Gray Ammo Box has only one visibly different layer at the thinnest point of only 6 walls.  You can't see it without a flashlight at a very steep angle.  This model was always printed with the lid and and insert tray, so the first few 10 to 12 take minutes.  Next 40% take about 40 seconds, and the last bit varies from 20 to 12 and back to 20.  So a wide swing.






Some I threw on a machine at work but didn't have time to send til tonight.  The PETG and ASA from Flashforge are amazing already, but with this build, these are the best prints I've ever gotten.  Some of the Voron parts look like injection molded parts from a few inches away.

One side effect is that with hot filaments, printing consistently like that helps the Z seams blend/disappear.  Irregularities in extrusion speed and cooling time, as you know, contribute to warping, and the Z seams gets the worst of it when at corners or small perimeters.  So this at least helps control one of those problems.  Yes, there are hours of tuning involved as well, but there's a Z seam on the corners of those boxes and I'm going to guess if I didn't tell you that, you would never have guessed it.

I'll take pics of the failures and other 2.1 beta build prints later.  But I'm heading to Michigan to look at adding robots on some rotary cncs.... in 5 hours.

Let me know if you're interested in printing anything else specific.  I have a ton of different ABS and ASA..

Have a good night!

**Comment by igiannakas** [comment]
This is fantastic. Ok I’m convinced I’ll put in some effort to turn it from POC to feature.

**Comment by igiannakas** [comment]
Posting some of my investigation results on the effects of layer time and speed below. I was working on the hypothesis that layer time also plays a role (big? Small?) on external wall consistency so decided to run some tests.

**Left equal layer time - Right equal speeds**


**"Equal" layer time:**
Layer time:
image

Speed used:
image

The layer time for each layer was between 43.1 sec and 44.1 sec in the blue areas after the screw stopped printing so a deviation of approx 2%. Where there is a solid layer (green area) the layer time was 47.2 sec, so a deviation of 9%.

**Equal speeds:**
Layer time:
image
Speed:
image


**Honestly I am puzzled. Or maybe I should be?**
1. **The equal layer time cube even though it was printed much slower overall (speeds of around 18mm/sec) turned out looking far worse than the equal speeds cube** (that was printed at 140 & 250 mm/sec external and internal perimeters). That I did not expect as I thought that printing slower and with less variation in layer time and speed it would turn out better. But nope...
2. It doesn't appear that the 2% variation in layer time in the top part of the cube is causing these print artefacts - they also dont seem to align up with the 2% variations in layer time from the print preview. But equally the print speed in those areas doesnt vary hugely. So there may be something else at play here.
3. **Where there is a sharp increase in layer time (the orange band in the middle of the print) in the equal speeds cube I see no or very minuscule artefacts**.
4. It probably looks like equal speeds between layers is the bigger lever to reduce slight cross layer variation but that is uncertain...
 
More experimentation is needed. I'm printing another cube with a slower perimeter (50mm/sec) and equal speeds to compare to the faster equal speeds cube. As I need to rule out that for some reason the Voron likes to print faster :P 

If it turns out that equal speeds matter the most then the original PR I've got open is probably the way to go - try to limit speed changes as much as possible caused by variations in slowdown due to layer time not being met.

**Comment by Zibit-Meowe** [comment]
Yeah, I'll definitely try a couple more ASA manufacturers.  I do have some from Polymaker here as well and they've looked about the same, slightly worse overhangs, but I have a lot of colors.  I'll try printing that Z Coupler in a few variations if I get time this week.

Heck, even just having a checkbox for "Prioritize surface finish" that did a std dev check or similar, and enabled it on parts where it had enough layer time to do so.

I also wonder if just segmenting out a consistent speed/fan setting only for external walls, or walls in general is useful.  As with everything, it depends on part size, an especially on the delta from tiny to large features.

Will try more when I'm back from road trip.

Thanks for your hard work on this!

**Comment by igiannakas** [comment]
So printed the exact same Gcode as last night for the fast print (150mm/sec) and consistent speeds and there is indeed some print to print variation
(the right two cubes - middle from last night, right from today, same gcode)
IMG_3867

Dont get me wrong, under normal viewing conditions you cannot see any of these minor artefacts but I think I'm fighting a loosing battle to be able to create a reproducible test to confirm or disprove my hypothesis above :(

Under regular viewing conditions:
image

**Comment by celtare21** [comment]
Leaving a comment here because this is a feature I really look towards and hope it becomes a reality.

From my experience, having a consistent outer wall speed does improve outer wall quality a lot, specially on different filaments. I mostly only print black and you can easily see any difference in shine when switching speeds. Depending on the filament manufacturer, this can be a small difference or look like a totally different filament, going from full shiny to matte. In this case, maintaining a constant speed is very important in look consistency and quality imo

**Comment by celtare21** [comment]
Also another scenario where this would improve quality a lot is when a specific speed creates a lot of VFAs. Using the CREATE_VIBRATIONS_PROFILE command from Klippain Shake&Tune, you can clearly see there's a lot of speed zones that make VFAs very visible, the best speed zone being the higher speed and the worst speed zones being the slower ones:
vibrations_20240604_005719
Being able to keep the speed high for the outer wall will minimize or totally remove the VFA artefacts, specially from shiny filament

**Comment by igiannakas** [comment]
Absolutely right on the ringing front. I've equally tuned my profiles to avoid those frequencies; however the sliver does hit them when slowing down for layer cooling which is annoying. I have also considered creating a "banned speeds" setting so anything that falls in a specific range is clamped either up or down. However the first step would be to allow more control on the external surfaces which this PR will do.

I've figured out the code fixes I needed to do and will push a good workable release for this in this PR today.

PS. your vibration graphs are pretty controlled anyway - I'm guessing these are 0.9 degree steppers? Your vibration is 1e7 vs 1e8 on mine below so 10 times lower!!

8CAA1AC5-9B9B-460D-BBF0-9B5BEE010C21

**Comment by celtare21** [comment]
> Absolutely right on the ringing front. I've equally tuned my profiles to avoid those frequencies; however the sliver does hit them when slowing down for layer cooling which is annoying. I have also considered creating a "banned speeds" setting so anything that falls in a specific range is clamped either up or down. However the first step would be to allow more control on the external surfaces which this PR will do.
> 
> I've figured out the code fixes I needed to do and will push a good workable release for this in this PR today.
> 
> PS. your vibration graphs are pretty controlled anyway from the looks of it, because you're running 0.9 degree steppers. Mine at 1.8 degrees are a bit worse (check the graph scale) but getting a pretty clean low end speed and med-high end speed. Hence why I test at 50mm/sec and 140mm/sec
> 
> 8CAA1AC5-9B9B-460D-BBF0-9B5BEE010C21

That would also be an awesome feature! The ability to define some speed zones the slicer would prefer and some banned speeds it should never hit, but yes I'd love to see this PR be implemented.
For me from the motion system, (running a custom ZeroG Mercury One) low speeds create awful VFAs that only go away while being at 160+mm/s. At 40mm/s you can clearly see artefacts through the print (my x rail is really bad and bindy, even tho hardware should be fixed before implements hacks in software, I believe this will also help people with some issues in their builds).

**Comment by igiannakas** [comment]
OK feature is ready and appears to work reliably in my tests. Tests performed:
1. Single extruder/Single filament in models with overhangs
2. Single extruder/single filament in models with overhangs and extrusion rate smoothing applied 
3. Multi material SEMM print with one filament having this option enabled and one having this option disabled. The enabled filament maintains a consistent outer wall speed vs. the disabled one that varies to meet layer time calculations

I think this is good to go now! For me, this option will now be permanently on for all my filaments.

@SoftFever all yours for code review etc :) and if you're feeling it for inclusion in the next beta round :P :D The code changes are not super invasive to say the least so not much should break (fingers crossed).

I've updated the first post summarising the feature etc.

**Comment by igiannakas** [comment]
@celtare21 @Ataraxia-Mechanica @ramosglauco @StuSerious @Zibit-Meowe the feature is now ready. thank you for your inputs and tests! Please feel free to download the latest PR build and take it through its paces. Please see the first post for the final feature description, where the option is etc.

**Comment by tastyratz** [comment]
You know, some of this might also be mitigated through somehow buffering the layer time slow down effect as well by means of a decay?

It seems like for now, layer speeds are modified to meet layer time requirements on an individual layer by layer level which creates the wild inconsistencies.

What if instead a layer time setting 'violation' is detected and a modifier gets applied with a multi layer effect? 
I.E. layer time violation detected, slow down layer by 50%, layer before and after that apply 50% of that 50% decay (25% speed reduction)  and 2 layers beyond that maybe 15% decay - obviously the rate of decay would have to be variable and somehow consider layer times that don't exist yet. Maybe layer time reductions are instead flagged instead of making per layer speed adjustments and then any slowdowns to meet layer time are handled in post processing by a second pass against flagged layers?

I'm spitballing and hypothesizing where I know it's significantly more complex than I make it sound. I'm mostly trying to think about approach.

Something else to consider here is that pressure advance is calibrated statically based on a specific speed/flow in Klipper, right?
If you were to calculate your pressure advance at 2 different speeds/flows and then dynamically scale it then you would see more consistency across speed variations (but Klipper doesn't let us do that natively). 
What if the root cause of the surface quality degradation when speed variances happen is actually a function of those slight changes in flow rate that were never calibrated for?

Compensating for that by making minimum layer time slowdowns a multi layer consideration would drastically smooth the end result.

We could even use a variance for comparison to reduce impact on small features/bridging whereas base layer time changes by more than x% discard any previous layer time slowdown flags.

**Comment by igiannakas** [comment]
That is what I tried experimenting with in the prints above. I limited layer time variations to 2% but the print came out comparatively awful. And I cannot explain that - I would have thought that it would benefit but it didn’t. The ones where speed is more consistent were more uniform, hence I’m hypothesising that layer time doesn’t matter as much. But it would be lovely if someone else also run these experiments to confirm or disprove this.

for PA I’ve thought multiple times of creating a PA curve in the slicer and applying the right pa for the feature. It’s trivial to do, but again do we do a double model for acceleration and speed or just acceleration. This needs empirical testing which I’ll get to at some point :)

this article here on PA was quite interesting: https://www.duet3d.com/blog/duet3d-research-nozzle-back-pressure-speed-and-pressure-advance#:~:text=Why%20does%20the%20pressure%20advance,extruder%20force%20at%20higher%20speeds.

**Comment by tastyratz** [comment]
Someone also went ahead and wrote pressure advance scaling into a macro.

https://www.reddit.com/r/klippers/comments/o98xes/pressure_advance_becomes_way_too_aggressive_when/h3d42ny/

I did experiment with this, but, I had trouble dialing it in and it made on the fly adjustments line by line so it completely blows out the logs (although it is fun to experiment with).

Could you also still have maximum flow rate limits at play here? A function of flow rate/pressure and heat put into the part vs kinematics itself?

This is especially relevant if you have a bambu hotend for example that has a soft knee on the flow rate limitations with a long lead up versus other hot ends that tend to perform more uniformly until their limits are reached.

**Comment by igiannakas** [comment]
I’ve got a better idea for PA. I’ve written it on discord but copying here for reference.

 What I’ve noticed is that the PA value needed is less the faster you print. See pictures of the below tests from left to right at 50-100-150 & 200 mm/sec with constant acceleration at 3k.
IMG_3874

The idea in this feature would be to run 4-5 PA tests for a filament and use the resulting values to calculate a best fit model that characterises PA behaviour against speed. The user would simply add the 4-5 Pa test optimal values and their corresponding speeds in a text box and orca would do the rest.

Then whenever the slicer calls for a speed change, that model is invoked and the appropriate extrapolated PA value is assigned.

For example:
8ECE2FEB-06D5-40C5-BF0F-7EB870B2CC48

In my above characterisation of the above PA values a cubic model provides the least error (deviation from the samples). We can then use that model to say if we want to print a feature at  120mm/sec calculate the required PA at 0.0326

This could be beneficial to characterise a filament and its behaviour according to speed. Then you have a PA model that can be used to extrapolate PA at any speed…

The more samples you provide, the more accurate the model - at minimum you’ll need 3 data points - one for the slowest, one for the fastest and one for the outer perimeter speeds. Then we can get a best fit function against these and extrapolate for other features, including slowdown for overhangs etc. 

I know there is a klipper macro that changes PA based on feature type. But that relies on calibrating PA values for the specific feature speed. And if you change the feature speed then the PA value is wrong. https://github.com/hadesz/v2blorgr-config/wiki/Dynamic-Pressure-Advance

So I think a better approach would be to characterise the behaviour of the filament using model fitting (linear, power, cubic etc) find the model with the least error and use that to project the PA values.

The below has a sample c++ code I’m working on but have yet to polish and integrate into orca: https://pastebin.com/WDtyrzME

**Comment by igiannakas** [comment]
Ps what @Zibit-Meowe saw was improvements with this PR. Not with smoothing of layer time across layers :)

**Comment by tastyratz** [comment]
Oh now that is very interesting! I'd be curious to see that make its way into a pull as well sometime.

I don't mean to distract from the original pull request with a whole other definitely interesting subject, it was more that I was thinking what if the differences were byproducts of speed and rooted in PA/flow rates. 
Zibit was reporting great results with your PR but it seemed like you weren't replicating the results on your test cubes to the level they were? Unless maybe I misread something.

**Comment by igiannakas** [comment]
> Oh now that is very interesting! I'd be curious to see that make its way into a pull as well sometime.
> 
> I don't mean to distract from the original pull request with a whole other definitely interesting subject, it was more that I was thinking what if the differences were byproducts of speed and rooted in PA/flow rates. Zibit was reporting great results with your PR but it seemed like you weren't replicating the results on your test cubes to the level they were? Unless maybe I misread something.

No I have replicated their result, absolutely. That’s why I finished this PR :). I was looking for further improvements and to validate what’s more important, layer time or speed. And speed seems to be more important which also matches the results that they got ;)

**Comment by igiannakas** [comment]
Yeah I know, I’m head deep in the rabbit hole right now haha! I’ve got a few ideas like this floating in my head.

1. PA profiling by speed and possibly acceleration (to be proven)
2. Speed variance smoothing over multiple layers (for the internal perimeters - external ones are now sorted based on this PR)
3. Forbidden speeds - to clamp speeds away from zones that cause VFAs
And more… 😝😝

**Comment by Zibit-Meowe** [comment]
Just a heads up, I promised more pics, but am stopping home just long enough to grab a bag.  Have to head across state for an emergency service call.  Be at the plant by 5:30 in the AM.  Ugh.  So may be the weekend until I have time enough at home to do it.

After testing, I think one thing that could make this even better would be a fan setting specifically for the outer wall, not tied to layer speed.  Then you could set the wall speed where VFA is not present.  Set a finish fan speed.  But still have strong internal walls.
Probably another whole test item on its own.

I need to test much larger parts as well.  I'd imagine there is some transition where layer time plays a bigger role than consistent extrusion.  Just haven't found it in medium [up to 100 x 150] sized parts.
I've got some large parts I made for J5 Axis on FANUC robots.  They catch pneumatic lines and wiring and don't allow them to wrap up and snag.  They're cones with a simple flair and a counterbore hole.  So the layer times get smaller the higher you go up the part.

**Comment by Ataraxia-Mechanica** [comment]
>     1. PA profiling by speed and possibly acceleration (to be proven)

Another thing that influences pressure advance is line width. Wider lines need less pressure advance. I sometimes use 0.6 or even 0.8mm outer walls with a 0.4mm nozzle on models that doesn't require sharp corners to improve overhangs.

**Comment by igiannakas** [comment]
> > ```
> > 1. PA profiling by speed and possibly acceleration (to be proven)
> > ```
> 
> Another thing that influences pressure advance is line width. Wider lines need less pressure advance. I sometimes use 0.6 or even 0.8mm outer walls with a 0.4mm nozzle on models that doesn't require sharp corners to improve overhangs.

Yeap agreed. Need to run the same test with different line widths and accelerations to see the effect.

from what I’ve been reading different accelerations shouldn’t impact PA much if you’ve tuned your PA smooth time to a low enough value. For line width I’d imagine there is a difference, but how much is to be proven.

**Comment by igiannakas** [comment]
> Just a heads up, I promised more pics, but am stopping home just long enough to grab a bag. Have to head across state for an emergency service call. Be at the plant by 5:30 in the AM. Ugh. So may be the weekend until I have time enough at home to do it.
> 
> After testing, I think one thing that could make this even better would be a fan setting specifically for the outer wall, not tied to layer speed. Then you could set the wall speed where VFA is not present. Set a finish fan speed. But still have strong internal walls. Probably another whole test item on its own.
> 
> I need to test much larger parts as well. I'd imagine there is some transition where layer time plays a bigger role than consistent extrusion. Just haven't found it in medium [up to 100 x 150] sized parts. I've got some large parts I made for J5 Axis on FANUC robots. They catch pneumatic lines and wiring and don't allow them to wrap up and snag. They're cones with a simple flair and a counterbore hole. So the layer times get smaller the higher you go up the part.

If you’re printing new parts, please download the latest version from the PR build. It’s now working correctly across scenarios like overhangs etc that the previous one would stumble.

**Comment by SoftFever** [comment]
> OK feature is ready and appears to work reliably in my tests. Tests performed:
> 
> 1. Single extruder/Single filament in models with overhangs
> 2. Single extruder/single filament in models with overhangs and extrusion rate smoothing applied
> 3. Multi material SEMM print with one filament having this option enabled and one having this option disabled. The enabled filament maintains a consistent outer wall speed vs. the disabled one that varies to meet layer time calculations
> 
> I think this is good to go now! For me, this option will now be permanently on for all my filaments.
> 
> @SoftFever all yours for code review etc :) and if you're feeling it for inclusion in the next beta round :P :D The code changes are not super invasive to say the least so not much should break (fingers crossed).
> 
> I've updated the first post summarising the feature etc.

Awesome work!
Unfortunately, the bottleneck is on my side. 
I don't have enough time to do the code review and test for version 2.1.0. We will have to wait for the next release.

**Comment by igiannakas** [comment]
lol no worries! I’ve got the dynamic PA feature also in the works so it will be a fun release 2.2 🎉

**Comment by ramorimdias** [comment]
This is genius! It seems obvious now that you think about it. I've always disabled slow down for cooling as it almost always leaves finish artifacts on the surface. Cannot wait to have this implemented!

**Comment by igiannakas** [comment]
As soft fever said this will be in the next release which will take a while. In the meantime feel free to use this PR build to have a play and provide testing feedback.

**Comment by ramorimdias** [comment]
Can you share a compiled version? i'm not too familiar with that... 😅

**Comment by Ataraxia-Mechanica** [comment]
Assuming you are using the windows version:
image
image

**Comment by Ataraxia-Mechanica** [comment]
Untitled
Other than this I think the option to disable outer wall slow down should disappear when "slow printing down for better layer cooling" is disabled?

**Comment by igiannakas** [comment]
Lol yeah... I'll fix it :)

**Comment by igiannakas** [comment]
> Untitled Other than this I think the option to disable outer wall slow down should disappear when "slow printing down for better layer cooling" is disabled?

Fixed

File: src/libslic3r/PrintConfig.cpp [hunk 1]
```diff
@@ -1700,6 +1700,15 @@ void PrintConfigDef::init_fff_params()
     def->tooltip = L("If enable this setting, part cooling fan will never be stoped and will run at least "
                      "at minimum speed to reduce the frequency of starting and stoping");
     def->set_default_value(new ConfigOptionBools { false });
+    
+    def = this->add("dont_slow_down_external_perimeters", coBools);
```
Comment (SoftFever) [inline src/libslic3r/PrintConfig.cpp:1704]: To make it shorter and consistent with other parameter, what do you think if we call it
`dont_slow_down_outer_wall`
and `Don't slow down outer walls`?

Comment (igiannakas) [reply src/libslic3r/PrintConfig.cpp]: Done :)

**Comment by Disttrack** [comment]
I love this feature so much, I've been waiting for it for years! I have done some tests and the result is amazing!
IMG_20240615_214514
IMG_20240615_214537

**Comment by SoftFever** [comment]
@igiannakas 
I managed to find some time to review and test it.
I believe the changes are safe to include in the coming RC version!
Thanks for you fantastic work!

**Review by SoftFever — APPROVED** [review]

**Comment by igiannakas** [comment]
Awesome 🎉😝

**Comment by Zibit-Meowe** [comment]
Well, best laid plans and all that.  First weekend I've been home since I posted last.

I did try printing the same objects again to reproduce the issue I referenced originally.  I'm certain it was filament now.
I could only get it with one roll of overture [ I swear, I have no luck with Overture ] and I can't reproduce it with good filament.

Included are pictures of the only reproducible [across filament] defects I had.  A better way to say it might be:  These are the defects _normally present_ that this doesn't solve.  It only really shows in PETG with a color shifting affect.  It's because of the minor shrinkage difference of the thin layer vs the thicker much slower layer below.  It's there in the others, but there's no polarizing/color effect in those to enhance it.

These are the worst examples.

With broad face lighting, it's barely visible as a color difference.


Flashforge Burnt Titanium PETG, brightly lit from steep angle above.


Same lighting.  Below is the overture crap.  Not the same issue, but you can see on this one the crazy issues on the seam.  Junk filament.  I tuned it eventually, but the Flashforge I had tuned in one print and prints with zhop at 3 times the speed.  Too much time to waste to save a few bucks a roll.


I was able to mostly get rid of these issues on the Flashforge material by tweaking fan and layer time.  But we're dealing with optical physics here.  You move that wall in 0.01 mm in or out and you will have some change in perceived color as the angle of incidence changes.  You can't feel the one on the right.

Speaking of which.  I really do think this is why speed works when it seems layer time would be better.  Now bear with me... it's amateur plastics scientist time!

You have a laminar flow of material which cools and crystalizes.  How fast and at what temp it is extruded seems to affect the final crystal structure.  It's likely related to how uniform the heating is.  Keep the temp and flow [ because speed ] constant and you'll get consistent heat transfer on long stretches like walls.  I'll ask one of the injection mold designers at work about it.

Oh and I still want an outer layer fan setting. ;D

Second issue I think is something unrelated to the change.  I'm having strange issues with small gap bridging on text, but haven't got to test vs 2.0 or 2.1 beta. I think it's related to the number of upper layers and some interaction with Ensure Vertical Shell Thickness.  When that isn't set to ALL, it goes nuts.  Here you can see the anchoring is only about  10% wall width.

Screenshot 2024-06-16 125738

Gotta see if there's an open issue for that.  Almost certainly not just this build.

Thanks everyone for your hard work!

**Comment by igiannakas** [comment]
Agree with you that speed is more of a factor compared to layer time- also because there is the factor of extrusion variation the faster you go. Typically the nozzle flow is not constant. Ie the faster you go the more you under extrude which can show as slightly thinner the extrusions, appearing like external wall artefacts. 

Layer time was an interesting theory but I’ve abandoned it as extrusion variations are more impacted by print speed (hence flow) as you’ve also mentioned and seen ;)

**Comment by igiannakas** [comment]
> I love this feature so much, I've been waiting for it for years! I have done some tests and the result is amazing! IMG_20240615_214514 IMG_20240615_214537

That is indeed a big difference!! 🎉 exactly the use case this PR was trying to solve.

**Comment by tastyratz** [comment]
> Speaking of which. I really do think this is why speed works when it seems layer time would be better. Now bear with me... it's amateur plastics scientist time!
> 
> You have a laminar flow of material which cools and crystalizes. How fast and at what temp it is extruded seems to affect the final crystal structure. It's likely related to how uniform the heating is. Keep the temp and flow [ because speed ] constant and you'll get consistent heat transfer on long stretches like walls. I'll ask one of the injection mold designers at work about it.


A bit off topic, but, this is something I've always thought about where concept of cooling and time/shift was thought of backwards when it comes to 3d printing.
When we know the volume of material being extruded and the temperature with which it's extruded we can probably calculate the amount of BTU being put into a printed part and better understand cooling if we can somehow calculate the material BTU, the heat transfer rate of a stationary nozzle (which applies BTU as a function of time over speed), and the amount of heat removed by cooling/fans while having a scalable offset considering chamber temperature making for more dynamic resilient prints. It's always bothered me that sliced files are assuming performance with a fixed chamber temperature and the slicer doesn't typically assume more chamber temperature scaling.
Some materials might hold more BTU, have a slower heat transfer/cooldown rate, etc. and that could be considered in material profiles. 

All of this is a calculation math/science well beyond my capacity. I don't know if I'm making sense or not.

**Comment by mikeq3** [comment]
This feature made a huge difference on my K1-Max prints. I had everything dialed in and still had a slight ringing on the outer walls. After this setting was enabled the ringing is gone.

**Comment by ramorimdias** [comment]
I have been getting failed prints when this feature is active for two reasons I think:
- in tall parts, at the end it prints too quick and even when slowing internal perimeters or infill, it cannot meet the minimum layer time
- in some prints, some small shells that only external perimeters

For example: I've tried to print twice the mouse 002 shell from bambulab and get failed at the same part even after brim added.



WhatsApp Image 2024-07-06 at 08 41 51

WhatsApp Image 2024-07-06 at 08 41 51 (1)


Could we think in a way to bypass this option when the minimum layer time is not respected? This would avoid some of my problems, I think

**Comment by igiannakas** [comment]
This failure is related to bed adhesion, not minimum layer time. 

The minimum layer time artefacts show as “melted” perimeters where they have normally solidified. In this case I don’t see any perimeters suffering from low cooling - they are shaped correctly.

Suggest a few of things that will help you:
1. Use hairspray or glue stick on the bed (I’m assuming it’s already clean, as in with a kitchen sponge and dish soap).
2. Enable slowdown for overhangs if it’s not enabled and make sure slowdown for curled edges is enabled too
3. Make sure your z hop is set up
4. Make sure the right bed plate is selected in the slicer. Bambu applies automatically lower z offset to the textured plate and if you select the smooth pei or cool plate it won’t so you will get adhesion issues

**Comment by ramorimdias** [comment]
Yeah, I know that improving bed adhesion can help with this print. The thing is that I've rerun the print with the default slowing down and it printed flawlessly. I'm not trying to accuse anything but I thought it could have room for improvement for some cases if you haven't though about that. 

I understand also that with such high speed external perimeters, the drag forces during extruding would be enough to tip the part off....

**Comment by igiannakas** [comment]
Yes of course, the reason it printed flawlessly when slowdown is enabled for all surfaces, is because the print was slower hence there were less forces on the part to knock it over.  

However what really needs to be addressed is the underlying reason for the failure.

If the external wall speeds are too fast for the nature of the model, the solution should be to reduce the external wall speeds in the profile to something that better matches the model, not create exceptions in this feature. 

Why I’m saying this? Imagine you printed 3-4 copies of this model on the same plate. Then the speeds on the top would not be slowed down due to minimum layer time, at all. Because you’re printing more copies, the others have time to cool down. But your model would still fail because the profile speed is too fast and the nozzle would knock it over anyway, even though minimum layer time is effectively inactive.

no offence taken at all :) I just think the approach suggested above is backwards. Enable this option and just reduce the wall speeds to 100mm/sec or so which would allow you to both have excellent surface finish, consistent shine and address the bed adhesion issue :)

**Comment by ramorimdias** [comment]
This makes completely sense. You are right!

**Comment by raphael008** [comment]
Hi @igiannakas 
It seems that we have discovered the same issue. I have already submitted a ticket to Bambulab in May. But they still put the feature in backlog I guess. Could you also submit a PR for this feature in Bambulab repo? Thank you so much for fixing this appearence issue.

**Comment by igiannakas** [comment]
> Hi @igiannakas It seems that we have discovered the same issue(I linked your issue in my ticket, please see above). I have already submitted a ticket to Bambulab in May. But they still put the appearence enhancement feature in backlog I guess. Could you also submit a PR for this feature in Bambulab repo? Thank you so much for fixing this appearence issue.

There are many factors that influence external print appearance, one of them being speed. This PR addresses the external wall speed only - and as such works best for shine variations. However as the internal perimeters are still printed slowly there will be or may be some deformation of the external perimeter if printing in inner outer mode.

id be curious to see your test results with this change using orca and report back if you can?

**Pull request merged**